### DISCUSSION TITLE
pyfluidysnth users?

### DISCUSSION THREAD

**Discussion by drtquin** [FluidSynth/fluidsynth#958] [Q&A]
Sadly I'm a Python programmer, not a C programmer.

I've made some good progress with a music teaching application based around pyfluidsynth, the Python Fluidsynth wrapper. However I now need to understand how in Python I can intercept events generated by a MIDI keyboard. 

Any pointers to helpful sources much appreciated.

**Comment by mawe42** [comment]
Hi!

It would be easier to help if you could explain a bit more what you want to achieve. There are a few different ways to intercept and modify MIDI events before they are rendered by FluidSynth. The easiest way might be either the FluidSynth-internal MIDI router or an external application (like mididings, for example).

Cheers
Marcus

**Reply by drtquin** [reply]
Hello Marcus,

Many thanks for taking the time to reply.

I'm writing in Python and using a modified Pyfluidsynth library to talk to Fluidsynth. My C programming skills are not sufficient to implement a C based call-back function.

I want to pick up the MIDI message from a MIDI keyboard, do stuff with it (for instance update a graphical musical display), then pass on a possibly modified message to Fluidsynth via Pyfluidsynth.

Conceptually I think I need to replace fluid_synth_handle_midi_event() or implement handle_midi_event_func_t in Python.

Mididings looks interesting, however I'm working on Windows 10, so I don't think I can use it.

Cheers,

¬Jonathan

**Comment by jimhen3ry** [comment]
I write in C++ for Windows and I think I do something similar to what I think you want to do. In my case I take MIDI In and manipulate it to simulate a theatre pipe organ, including making a graphical display of the organ console that shows what the MIDI In is doing. I only use FluidSynth for the output. You can use FluidSynth's MIDI Channel Messages, such as fluid_synth_noteon, pretty much as a one-to-one replacement for MIDI Out messages. I have all the FluidSynth output paralleling Windows MIDI Out messages so that my program can be configured to output either MIDI or FluidSynth audio.

**Reply by drtquin** [reply]
Does sound similar, except that you're able to program in C++, whereas my skills are really only sufficient for Python. I've got around the problem by finding out how to stop Pyfluidsynth grabbing MIDI input, which then allows me to use Mido for MIDI input.

**Reply by jimhen3ry** [reply]
It sounds like your workaround is about the same as what I was suggesting. Use FluidSynth only for output. Stopping Pyfluidsynth grabbing MIDI input sounds more like a solution than a workaround.

**Comment by albedozero** [comment]
Hi @drtquin - I've contributed a decent amount to pyfluidsynth, and I've written my own Python FluidSynth wrapper as part of another project that implements a [custom midi message router](https://github.com/albedozero/fluidpatcher/blob/8055dbc01849bd1dc9bf502f8e0cbf2d58633862/patcher/fswrap.py#L557) that I think does the kind of thing you're looking for. If you look at the linked function, as well as the `Synth.__init__` function you should be able to see how the callback is set, but feel free to ask me questions if you're interested! I need to comment my code better...

**Comment by drtquin** [comment]
Hello Albedo Zero,

Excellent to hear from you. This is robust and professional work, but I think I'll need a bit of  help to use it ;o(

- Is fswrap.py a replacement for pyfluidsynth? pyfluidsynth provides simple access to things like noteon and noteoff. I couldn't see equivalent calls in fswrap
- Can fswrap.py be used by itself, or does it need other packages in the fluidpatcher-0.4.2 directory?
- I added a couple of methods plus wrappers to pyfluidsynth to allow me to do some tuning to experiment with other than 12 tones per octave (eg 17 and 19):
-- fluid_synth_activate_key_tuning
-- fluid_synth_activate_tuning
-- fluid_synth_deactivate_tuning
-- I tried and failed to implement fluid_synth_tuning_dump
- Does fswrap provide access to the above tuning methods?
- What is the "linked function" you mentioned
- Might you have some basic example code?

Many thanks indeed.

**Reply by albedozero** [reply]
Sure thing! I shared fswrap.py with you mainly so you could see how to create your own custom router function in Python, maybe in the fork of pyfluidsynth you've worked on. fswrap.py _is_ a replacement for pyfluidsynth (mainly for live playing), and could be used by itself, but its API is a little bit different (no wrappers needed for note_on, note_off calls, for example, since it assumes a MIDI keyboard will be attached). Up to you whether to adapt it or modify your own fork of pyfluidsynth.

To create your own midi router function in Python, it should accept two arguments (like my [custom_midi_router](https://github.com/albedozero/fluidpatcher/blob/8055dbc01849bd1dc9bf502f8e0cbf2d58633862/patcher/fswrap.py#L557) function - that's what I meant by the "linked function" in my last post). For FluidSynth's MIDI driver to be able to talk to your Python function, it should be declared as a C callback function type, which you can do with the handy CFUNCTYPE function (which lets you set the proper argument and return types).
```python
custom_router_callback = CFUNCTYPE(c_void_p, c_void_p, c_void_p)(custom_midi_router)
```
Then you need to pass your C callback as the second argument to _new_fluid_midi_driver_ (which is called [here](https://github.com/albedozero/fluidpatcher/blob/8055dbc01849bd1dc9bf502f8e0cbf2d58633862/patcher/fswrap.py#L543) in fswrap or [here](https://github.com/nwhitehead/pyfluidsynth/blob/dac7f04bfe4cafc505a0abe35543503b9c8cc602/fluidsynth.py#L592) in pyfluidsynth).
```python
new_fluid_midi_driver(settings, custom_router_callback, destination)
```
The _destination_ argument is a function that gets passed to your custom router function so you can send the messages somewhere. Usually this is _fluid_midi_router_handle_midi_event_, so FluidSynth can do its own handling of MIDI events, but you could also send them directly to the synth created by _new_fluid_synth_. To actually manipulate the incoming MIDI events, you may find the _[MidiEvent](https://github.com/albedozero/fluidpatcher/blob/8055dbc01849bd1dc9bf502f8e0cbf2d58633862/patcher/fswrap.py#L159)_ class in fswrap a useful example.

I realize this is a complicated and yet incomplete explanation, but I didn't want to go on too long and thought it would give you enough to go a bit further. Feel free to ask more.

Devs, please speak up if this isn't relevant enough/hijacking and we should move the discussion to an issue on one of these other mentioned repos, or perhaps to @drtquin's fork. Could always drop a link here.

**Reply by drtquin** [reply]
Hello there!

This line in fluidsynth.py:
`self.midi_driver = new_fluid_midi_driver(self.settings, self.custom_router_callback, self.router)`

is generating the error:
`ctypes.ArgumentError: argument 2: <class 'TypeError'>: expected CFunctionType instance instead of CFunctionType`

I've tried variations, both give the same error. 

The first follows fluidpatcher.py, so I've added this statement near the start of the file:
`fl_eventcallback = CFUNCTYPE(c_void_p, c_void_p, c_void_p)`

and at the end of Synth.start() I have:

`
self.custom_router_callback = fl_eventcallback(self.custom_midi_router)
self.midi_driver = new_fluid_midi_driver(self.settings, self.custom_router_callback, self.router)
`

The second attempts to follow your email, so at the end of Synth.start() I have:
`
self.custom_router_callback = CFUNCTYPE(c_void_p, c_void_p, c_void_p)(self.custom_midi_router)
self.midi_driver = new_fluid_midi_driver(self.settings, self.custom_router_callback, self.router)
`

I've added a slight variant of your `custom_midi_router()` to the Synth class.

Very many thanks for any help you can provide.

**Reply by albedozero** [reply]
Either way should work - it was actually a bug in pyfluidsynth. The new_fluid_midi_driver function arguments were specified incorrectly, so your callback function couldn't match up - sometimes those ctypes errors are a bit cryptic. I updated the pyfluidsynth repo, but if you modify the same line in your code [like this](https://github.com/nwhitehead/pyfluidsynth/commit/00d19470c2fc10f2c38f470abaa18ba62c76e1c6) it should work for you.

**Reply by drtquin** [reply]
Hello Bill,

Almost there. With the change to `new_fluid_midi_driver ` I can now get my own handler to work, so I can successfully intercept MIDI keyboard events and do my own thing with them.

I've added a parameter to `fluidsynth.synth.Start()` to allow me to pass in my custom handler: If I pass in "None" I want to revert to the default handler.

However the change `new_fluid_midi_driver` causes this line (from the original Start() method)
`self.midi_driver = new_fluid_midi_driver(self.settings, fluid_midi_router_handle_midi_event, self.router)`
to generate the error:
`ctypes.ArgumentError: argument 2: <class 'TypeError'>: expected CFunctionType instance instead of CFunctionType`

**Reply by albedozero** [reply]
Oops - I had the return types for the `new_fluid_midi_driver` and `new_fluid_midi_router` callbacks defined incorrectly. That first fix wasn't quite right, but if you make [these changes](https://github.com/nwhitehead/pyfluidsynth/commit/8b03b7c1248550468754d55f5809a14d7a8c739d) in your _fluidsynth.py_ it should work correctly. Also, make sure your custom router function returns FLUID_OK or FLUID_FAILED. Thanks for helping me track down my own 4-year old bug!

**Reply by drtquin** [reply]
Ah, the sweet sound of success! Very many thanks for your help. 

All works but some rather ugly code: 

`fluidsynth.synth.Start()` now looks like this:
```
    def start(self, driver=None, device=None, midi_driver=None, midi_router=None):
         <stuff>
         if midi_router == None:
            self.midi_driver = new_fluid_midi_driver(self.settings, fluid_midi_router_handle_midi_event, self.router)
        else:
            self.custom_router_callback = CFUNCTYPE(c_int, c_void_p, c_void_p)(midi_router)
            self.midi_driver = new_fluid_midi_driver(self.settings, self.custom_router_callback, self.router)
```
I've had to add half a dozen getter methods in `pyfluidsynth` to  get the data from a MIDI event, so my `midi_router` looks like this:

```
def custom_midi_router(self, data, event):
        MIDI_NOTEOFF = 0x80
        MIDI_NOTEON = 0x90
        MIDI_CONTROL = 0xb0
        MIDI_VOLUME = 0x07
        type = self.fs.midi_event_get_type(event)
        velocity = 0
        key = 0
        value = 0
        if type == MIDI_NOTEOFF:
            midiKey = self.fs.midi_event_get_key(event) 
            self.fs.noteoff(0, midiKey)
            <stuff>
        elif type == MIDI_NOTEON:
            velocity = self.fs.midi_event_get_velocity(event)
            midiKey = self.fs.midi_event_get_key(event)
            self.fs.noteon(0, midiKey, velocity)
            <stuff>
        elif type == MIDI_CONTROL:
            control = self.fs.midi_event_get_control(event)
            value = self.fs.midi_event_get_value(event)
            <stuff>

        return fluidsynth.FLUID_OK
```

Thanks once again.

**Reply by albedozero** [reply]
Hey, whatever works! Happy to help. Make some awesome music!

**Comment by drtquin** [comment]
Hello Bill,

 

Can I test your patience once again?

 

Having successfully intercepted messages from a MIDI keyboard, I now want to do the same from a file.

 

I can load and play a file using new_fluid_player(), fluid_player_add(), and fluid_player_play(), however rather than have fluidsynth play the file directly, I want to intercept the MIDI events.

 

I think I can do this using the custom router call back we created for intercepting MIDI keyboard events. I’ve tried using this when creating the router, ie 

self.router = new_fluid_midi_router(self.settings, self.custom_router_callback, self.synth)

However this seems to do nothing: it doesn’t crash and the player still plays the file.

 

Guidance appreciated!

 

A different question: I think I’ve broken something somewhere. When I play a note on my physical MIDI keyboard, or my virtual screen keyboard, I get a very short note. This is the case for both percussive (eg piano) and sustained (eg organ) instruments. Any thoughts?

 

Very many thanks once again.

 

¬Jonathan

**Reply by albedozero** [reply]
Hi, Jonathan - no problem. I'll answer the second question first by saying I think you just need to grab the latest release of FluidSynth - the note-squashing sounds like #935, which was fixed in v2.2.2.

To get your player to send its events to your custom (or any) router, you just need to call `fluid_player_set_playback_callback()` on it, e.g.

```
fluid_player_set_playback_callback(self.player, self.custom_router_callback, self.synth)
```

When first created, the player's playback_callback is set to send directly to the synth by default, which is why you can hear it but your router has no effect.

**Comment by drtquin** [comment]
Hello Bill,

 

Fantastic!

 

1.	The latest download (2.2.3) did indeed solve the short note problem. I thought I’d broken something!
2.	Your arguments for fluid_player_set_playback_callback(self.player, self.custom_router_callback, self.synth) did the trick. I’d started down this road, but got stuck because the third argument is different to that given at https://www.fluidsynth.org/api/group__midi__player.html#ga615cfa9d81227d9253dddebc49d8edcb

 

I’m happy to share the work I (we) have done on fluidsynth.py. Is there an interest group or similar who might be interested?

 

Very best,

 

¬Jonathan

**Reply by albedozero** [reply]
Sharing is very welcome! I think the best place to share would be on the [main pyfluidsynth repo](https://github.com/nwhitehead/pyfluidsynth) - that's where most of the (sporadic) development seems to be going on. You could start a discussion there, but the much slicker (and easier to manage) way would be to click the little fork button in the upper right-hand corner of the page, then copy and paste your changes into your own (now-forked) version of the pyfluidsynth repository. Github will then probably suggest you make a pull request, or you can click "contribute", and that request will show up in the main repo and we can do all our discussion/commenting/updating there.

**Reply by drtquin** [reply]
Shared as you suggest. Not sure if I've done the submission correctly. Thanks again.

**Comment by drtquin** [comment]
I'm trying to stop a file playing by calling `fluid_player_stop()` followed by `delete_fluid_player()`. However if there are notes playing when these methods are called, the notes continue to play. Which is a real problem when playing an organ piece!

**Reply by albedozero** [reply]
I noticed this recently too - if sustain or sostenuto are active when you stop/pause, the notes keep going, unless you also call `fluid_player_seek(0)`. I've been discussing this with the maintainers in #981 and it might get simpler if my changes in #980 are added.

**Comment by drtquin** [comment]
Per the above, I'm using the fluid player to read in a MIDI file, and intercepting the events generated by the player.

I have a handle on most events being generated, however I'm not sure how to handle event 0x51. I think this means set tempo, however I can't see how to extract the tempo from the event.

Once I have the tempo, I assume I can set the tempo with fluid_player_set_tempo()

**Reply by albedozero** [reply]
It is a tempo change - you can extract it using any of the functions that get param1. For example `fluid_midi_event_get_key`, `fluid_midi_event_get_velocity` - there are several functions related to specific event types, but all they do is return the param1 member of the event struct. This is probably so some event-specific functionality could be included in future, but I'm not sure what that would be and I suspect it won't happen.

If you're not doing anything to modify an event/event type, you could just pass it on to fluidsynth's router by calling `fluid_midi_router_handle_midi_event()` at the end of your custom router function - no need to rewrite your own code to handle everything.

**Reply by drtquin** [reply]
Is this correct? Looking at https://www.recordingblogs.com/wiki/midi-set-tempo-meta-message suggests that the set tempo message has 3 bytes of data, not 1.

Also the value returned by eg `fluid_midi_event_get_velocity()` or `fluid_midi_event_get_value()` is 0.

With thanks.

**Reply by albedozero** [reply]
Oops my bad - `fluid_midi_event_get_velocity` and `fluid_midi_event_get_value` both return param2, so you would want to use e.g. `fluid_midi_event_get_key`. Tempo messages are indeed 3 bytes, but fluidsynth converts those bytes into a single integer for you and stores them in param1 [here](https://github.com/FluidSynth/fluidsynth/blob/926581851ed1a095ef5b8659f77b38272d57e624/src/midi/fluid_midi.c#L875).

**Reply by drtquin** [reply]
Once again many thanks for your help. 

A typical tempo value I get back is 0x35dac2. I've added a `fluid_player_set_tempo()` wrapper to fluidsynth.py, and I feed the tempo into this. 

This method also takes a `tempo_type` parameter which can be one of
```
FLUID_PLAYER_TEMPO_INTERNAL 
FLUID_PLAYER_TEMPO_EXTERNAL_BPM 
FLUID_PLAYER_TEMPO_EXTERNAL_MIDI
```
I'm guessing these translate to integers 0, 1, 2, respectively (and that I want the third option), however it would be nicer to import these values from fluidsynth into fluidysnth.py.

Yet again my software ignorance: how can I do this?

**Reply by albedozero** [reply]
Values like that are defined as _enum_s in the C code, for example [here](https://github.com/FluidSynth/fluidsynth/blob/a0ec2c59326b311a7cbaeb638c2002cf2c8c82db/include/fluidsynth/midi.h#L257) - if values aren't specified for the constants they just get assigned 0, 1, 2, etc. so your assumption is correct. I don't think there's a way to export the values from the C code into python - easiest thing is just to define them yourself, for example

```python
FLUID_PLAYER_TEMPO_INTERNAL = 0
FLUID_PLAYER_TEMPO_EXTERNAL_MIDI = 2
```

Here, `INTERNAL` means "use the tempo messages in the MIDI file", `EXTERNAL` means "change the tempo to a specific value I'm giving you". The first is the default, and fluidsynth will handle all the tempo updates - you don't have to parse the messages and call `fluid_player_set_tempo` yourself, unless you're trying to modify the tempo events. Your custom router function only needs to do things to the events you want to modify - the rest you can just pass on to `fluid_midi_router_handle_midi_event` for regular routing, or directly to `fluid_synth_handle_midi_event` and let fluidsynth do its thing.

**Comment by jbigdata-git** [comment]
Hi,

Working on [DrumNotes](https://github.com/jbigdata-git/music21drums) an addon to [Music21](https://web.mit.edu/music21/) to play drum parts, I dev a _StreamPlayer_ full based on **FluidSynth**.
Among tests, **[this one](https://github.com/jbigdata-git/music21drums/blob/main/music21drums/fluidsynth_test.py)** do the job expected. 
I use **pyFluidSynth** and achieve the adds to use _FluidSynth 2.2.x_, more efficient to callback MIDI events.

_python3 fluidsynth_test.py_ and see 2 different callbacks to intercept MIDI events working.
_...
playback_callback [140674764536352, 140674137616464, <class 'int'>].
playback_callback [140674764536352, 140674137616512, <class 'int'>].
...
tick_callback [140571333214544, 37880].
tick_callback [140571333214544, 37882].
..._

Hope it can help.

JB.
